string.c

```diff
@@ -11856,18 +11782,131 @@ rb_enc_interned_str_cstr(const char *ptr, rb_encoding *enc)
 }
 
 /*
- *  A String object holds and manipulates an arbitrary sequence of
- *  bytes, typically representing text or binary data. String objects may be
- *  created using String::new or as literals.
+ *  A \String object has an arbitrary sequence of bytes,
+ *  typically representing text or binary data.
+ *  A \String objects may be created using String::new or as literals.
  *
  *  String objects differ from Symbol objects in that Symbol objects are
- *  designed to be used as identifiers, instead of text or data.
+ *  commonly used as identifiers, instead of text or data.
+ *
+ *  Some \String methods modify +self+.
+ *  Typically, a method whose name ends with <tt>!</tt> modifies +self+
+ *  and returns +self+;
+ *  often a similarly named method (without the <tt>!</tt>)
+ *  returns a new string.
```

Comment:
In general, if there exist both bang and non-bang version of methods, the bang! mutates and the non-bang! does not.  However, if a method without a bang can also mutate, such as `replace`.